### ISSUE TITLE
torch.linalg.inv breaks down when used on mps.

### ISSUE THREAD

**Issue by frantisekdracek** [pytorch/pytorch#165850]
### 🐛 Describe the bug

I have a code that is supposed to compute rolling multiple linear regression in pytorch. When I run this code on cpu, the results are very close to results from numpy.linalg.lstsq. However when I run the code on MPS, the result is completely wrong. 

example for 4 dimensional input
```
numpy: [1.00035739, 0.29986116, 0.19779432, 0.46009004]
cpu: [1.00035691, 0.29986125, 0.19779432, 0.46008980]
mps: [ 2.56928920e+07,  4.40364003e+14, -4.40364037e+14,  1.34075220e+07]
```

Here is a snippet of relevant part of function:

`
    # Reshape for batch matrix multiplication
    # x_conv: [B, D, N] -> [B, N, D, 1]
    x_expanded = (x*mask).unsqueeze(3)  # [B, N, D, 1]
    x_expanded_T = x_expanded.transpose(2, 3)            # [B, N, 1, D]

    # Batch matrix multiply: [B, N, D, 1] @ [B, N, 1, D] -> [B, N, D, D]
    # only last two dims are matrix-multiplied
    XTX = torch.matmul(x_expanded, x_expanded_T)      # [B, N, D, D]

    #we need to convolve each element of [D, D] matrix with fitler
    # Reshape XTX from [B, N, D, D] to [B, D*D, N] for convolution
    # Each element of the D×D matrix becomes a separate channel
    XTX_flat = XTX.permute(0, 2, 3, 1).reshape(B, D * D, N)  # [B, D*D, N]

    # Apply convolution to sum over windows
    XTX_conv = F.conv1d(XTX_flat, filters_XTX, groups=1)  # [B, D*D, out_len]

    # Reshape back to [B, out_len, D, D]
    XTX_conv = XTX_conv.reshape(B, D, D, out_len).permute(0, 3, 1, 2)  # [B, out_len, D, D]

    #inverse of x.T x
    XTX_inv = torch.linalg.inv(XTX_conv)                       # [B, out_len, D, D]
`

I have traced the function, and while XTX_conv has mismatch between cpu and mps on order of 1e-7, the computation completely breaks down when computing inverse. The matrix that is being inverted has full rank, but the computation seems to explode on mps.

### Versions

Collecting environment information...
PyTorch version: 2.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==2.3.1
[pip3] optree==0.16.0
[pip3] torch==2.7.1
[pip3] torchmetrics==1.8.2
[pip3] torchvision==0.22.0
[pip3] torchvision-extra-decoders==0.0.2
[conda] libtorch                  2.7.1           cpu_generic_hd8f8d4b_3    conda-forge
[conda] nomkl                     1.0                  h5ca1d4c_0    conda-forge
[conda] numpy                     2.3.1           py313h674b998_1    conda-forge
[conda] optree                    0.16.0          py313h0ebd0e5_0    conda-forge
[conda] pytorch                   2.7.1           cpu_generic_py313_heefb1e6_3    conda-forge
[conda] torchmetrics              1.8.2              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.22.0          cpu_py313_hc491b9c_1    conda-forge
[conda] torchvision-extra-decoders 0.0.2           py313hc37fbc8_3    conda-forge

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen

**Comment by inventshah** [comment]
I can't reproduce as of torch `2.8.0` (I suspect this was fixed from #146754). Are you able to upgrade torch? If not, the issue appears to be with non-contiguous data, so adding `XTX_conv.contiguous()` would fix the cpu-mps discrepancy in `2.7.1`.

A minimal repro might be:
```python
t = torch.tensor([[1.0, 2.0], [3.0, 4.0]]).permute(1, 0)
print(torch.linalg.inv(t))
print(torch.linalg.inv(t.to('mps')))
```
without `permute` the two results are the same.

**Comment by frantisekdracek** [comment]
The minimal repro behaves as you wrote; the result differs between CPU and MPS, and works as expected after upgrading to 2.8.0.
With regards to my original problem, after upgrading PyTorch and applying `contiguous()`, the results of `inv` between CPU and MPS agree, albeit with very low precision ~ 1e-3.

**Comment by inventshah** [comment]
Do you have a sample input (either `XTX_conv` or both `x_expanded` and `filters_XTX`)?

Using `torch.randn` for `x_expanded` and `filter_XTX`, I found the error between cpu-mps on average low. One example with large absolute error:
```python
t = torch.tensor(
    [
        [0.89, 0.02, 2.29, 1.11],
        [0.92, -0.25, 1.08, -0.10],
        [-0.58, 0.85, 1.52, 0.05],
        [0.44, -0.09, 0.27, -1.05],
    ]
)
r1 = torch.linalg.inv(t)
r2 = torch.linalg.inv(t.to("mps")).cpu()
print(torch.allclose(r1, r2))  # False
print(torch.aminmax(torch.abs(r1 - r2)))  # min=0.0305, max=1.2085
```
so definitely some difference, but comparing the cpu result with `numpy.linalg.inv` also has a large difference (`min=0.0359, max=1.4194`)...
